src/main/kotlin/org/rust/ide/inspections/RsDeprecationInspection.kt

```diff
@@ -0,0 +1,99 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.inspections
+
+import com.intellij.codeInspection.ProblemHighlightType
+import com.intellij.codeInspection.ProblemsHolder
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiElementVisitor
+import com.intellij.psi.util.parents
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+
+class RsDeprecationInspection : RsLocalInspectionTool() {
+    override fun getDisplayName() = "Deprecated item"
+
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor = object : RsVisitor() {
+        override fun visitElement(ref: RsElement) {
+            // item is non-inline module declaration or not reference element
+            if (ref is RsModDeclItem || ref !is RsWeakReferenceElement) return
+
+            var original = ref.reference?.resolve() ?: return
+
+            val identifier = ref.referenceNameElement ?: ref
+            original = when (original) {
+                is RsFile -> original.declaration
+                is RsAbstractable -> original.superItem
+                else -> original
+            } ?: original
+
+            if (original is RsOuterAttributeOwner && ref.parents().none { it.hasAllowDeprecatedAttribute() }) {
+                checkAndRegisterAsDeprecated(identifier, original, holder)
+            }
+        }
+    }
+
+    private fun checkAndRegisterAsDeprecated(identifier: PsiElement, original: RsOuterAttributeOwner, holder: ProblemsHolder) {
+        original.outerAttrList
+            .mapNotNull { DeprecatedAttribute.from(it) }
+            .firstOrNull()
+            ?.let {
+                holder.registerProblem(identifier, it.getMessage(identifier.text), ProblemHighlightType.LIKE_DEPRECATED)
+            }
+    }
+
+    private fun PsiElement.hasAllowDeprecatedAttribute(): Boolean = this is RsOuterAttributeOwner &&
```

Comment:
Let's move this code into `org.rust.lang.core.psi.ext.QueryAttributes`. This class contains some usefull methods

Reply:
I just replaced this method with a general analog the method from `org.rust.lang.core.psi.ext.QueryAttributes`